lib/ClangImporter/ImportDecl.cpp

```diff
@@ -2486,14 +2486,114 @@ namespace {
+        clang::SwiftAttrAttr *returnsRetainedAttr =
+            clang::SwiftAttrAttr::Create(clangCtx, "returns_retained");
+        generatedCxxMethodDecl->addAttr(returnsRetainedAttr);
+        generatedCxxMethodDecl->setAccess(clang::AccessSpecifier::AS_public);
+
+        // Getting the default constructor
+        clang::CXXConstructorDecl *ctorDecl = nullptr;
+        for (clang::CXXConstructorDecl *ctor : cxxRecordDecl->ctors()) {
+          if (ctor->isDefaultConstructor() && !ctor->isDeleted()) {
+            ctorDecl = ctor;
+            break;
+          }
+        }
+
+        if (ctorDecl) {
+          // Build the constructor expression using the default constructor decl
+          clang::ExprResult generatedConstructExpr =
+              Impl.getClangSema().BuildCXXConstructExpr(
+                  clang::SourceLocation(), cxxRecordTy, ctorDecl, false,
+                  clang::MultiExprArg(), false, false, false,
+                  true, // TODO: fix the propagation of this zeroinit flag to
+                        // BuildCXXNew
+                  clang::CXXConstructionKind::Complete, clang::SourceRange());
+
+          if (!generatedConstructExpr.isInvalid()) {
+            // Build the new expr by passing the constructor expression as an
+            // initializer
+            clang::ExprResult generatedNewExprResult =
+                Impl.getClangSema().BuildCXXNew(
+                    clang::SourceRange(), false, clang::SourceLocation(), {},
+                    clang::SourceLocation(), clang::SourceRange(), cxxRecordTy,
+                    clangCtx.getTrivialTypeSourceInfo(cxxRecordTy),
+                    std::nullopt, clang::SourceRange(),
+                    generatedConstructExpr.get());
+
+            if (!generatedNewExprResult.isInvalid()) {
+              // synthesizing the body of the generatedCxxMethodDecl
+              clang::CXXNewExpr *generatedNewExpr =
+                  cast<clang::CXXNewExpr>(generatedNewExprResult.get());
+              clang::ReturnStmt *generatedRetStmt = clang::ReturnStmt::Create(
+                  clangCtx, clang::SourceLocation(), generatedNewExpr, nullptr);
+              clang::CompoundStmt *generatedFuncBody =
+                  clang::CompoundStmt::Create(
+                      clangCtx, {generatedRetStmt}, clang::FPOptionsOverride(),
+                      clang::SourceLocation(), clang::SourceLocation());
+              generatedCxxMethodDecl->setBody(generatedFuncBody);
+
+              // Getting the appropriate decl context
+              DeclBaseName initBaseName = DeclBaseName::createConstructor();
+              DeclName importedInitName(ctx, initBaseName,
+                                        llvm::ArrayRef<Identifier>{});
+              EffectiveClangContext effectiveContext(
+                  generatedCxxMethodDecl->getDeclContext()->getRedeclContext());
+              auto dc = Impl.importDeclContextOf(decl, effectiveContext);
```

Comment:
We already have a similar code in this function above:
```
auto dc =
          Impl.importDeclContextOf(decl, importedName.getEffectiveContext());
```

Why not reuse the `dc` from there?

Reply:
But we might not even need this once we move to `importFunctionDecl`.